Segmentation fault after upgrade to 0.17 on macOS 11.6.2

After upgrading to IWYU 0.17 on my BigSur installation (within Homebrew), IWYU stopped working on my codebase due to some pesky segmentation fault.

By trial and error on one file that was segmenting I was able to reduce to a small snippet that cause the segmentation fault:

````
#include <algorithm>
#include <cassert>
#include <vector>

int main()
{
    constexpr auto unused = static_cast<std::size_t>(-1);
    const auto Ndof = 100ul;

    std::vector<std::size_t> mapping(Ndof, unused);

    assert(std::none_of(mapping.cbegin(),
                        mapping.cend(),
                        [](auto value)
                    {
                        return value == unused;
                    }));

    return EXIT_SUCCESS;
}
````

This  segmentation fault is highly specific:

- If the algorithm is not inside the `assert` command, no segmentation fault.
- If I declare the lambda as a variable `auto is_unused = [](auto value) {... } ` and then use `is_unused` in the `std::none` algorithm, no segmentation fault.

Before the upgrade to newest IWYU (LLVM was also upgraded - I did it through `brew upgrade`), IWYU was working on my entire codebase without issue.

I concur it is a serious problem, but I must admit I am a bit at a loss to provide you more...

I build the snippet with the following CMakeLists.txt:

````
cmake_minimum_required(VERSION 3.12)

project(Debug_IWYU)

set(CMAKE_CXX_STANDARD 20 CACHE STRING "C++ standard; at least 20 is expected.")
set(CMAKE_C_COMPILER /usr/local/Cellar/llvm/13.0.0_2/bin/clang)
set(CMAKE_CXX_COMPILER /usr/local/Cellar/llvm/13.0.0_2/bin/clang++)
set(CMAKE_CXX_INCLUDE_WHAT_YOU_USE /usr/local/Cellar/include-what-you-use/0.17/libexec/bin/include-what-you-use;-Xiwyu;any;-Xiwyu;iwyu;-Xiwyu;args;-Xiwyu;--no_comments;-Xiwyu;--cxx17ns;-Xiwyu;max_line_length=120)

add_executable(test
               snippet.cpp)   
````

and then the usual

````
mkdir build
cd build
cmake ..
````
(I have reduced to the minimum the fluff such as using Ninja rather than make and introducing a mapping file).

When I run `make`, I get a segmentation fault and no more information; as I said it's by trial and error I pinpointed what could make it work or not. If I run `make --trace` I don't get much more:

````
Makefile:179: target 'cmake_check_build_system' does not exist
/usr/local/Cellar/cmake/3.22.2/bin/cmake -S/Users/sebastien/Codes/Sandbox/IWYU -B/Users/sebastien/Codes/Sandbox/IWYU/build_iwyu --check-build-system CMakeFiles/Makefile.cmake 0
Makefile:90: update target 'all' due to: cmake_check_build_system
/usr/local/Cellar/cmake/3.22.2/bin/cmake -E cmake_progress_start /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu/CMakeFiles /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu//CMakeFiles/progress.marks
make -s -f CMakeFiles/Makefile2 all
CMakeFiles/Makefile2:82: target 'CMakeFiles/test.dir/all' does not exist
make -s -f CMakeFiles/test.dir/build.make CMakeFiles/test.dir/depend
CMakeFiles/test.dir/build.make:108: target 'CMakeFiles/test.dir/depend' does not exist
cd /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu && /usr/local/Cellar/cmake/3.22.2/bin/cmake -E cmake_depends "Unix Makefiles" /Users/sebastien/Codes/Sandbox/IWYU /Users/sebastien/Codes/Sandbox/IWYU /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu /Users/sebastien/Codes/Sandbox/IWYU/build_iwyu/CMakeFiles/test.dir/DependInfo.cmake --color=
Consolidate compiler generated dependencies of target test
make -s -f CMakeFiles/test.dir/build.make CMakeFiles/test.dir/build
CMakeFiles/test.dir/build.make:75: update target 'CMakeFiles/test.dir/snippet.cpp.o' due to: ../snippet.cpp
/usr/local/Cellar/cmake/3.22.2/bin/cmake -E cmake_echo_color --switch= --green --progress-dir=/Users/sebastien/Codes/Sandbox/IWYU/build_iwyu/CMakeFiles --progress-num=1 "Building CXX object CMakeFiles/test.dir/snippet.cpp.o"
[ 50%] Building CXX object CMakeFiles/test.dir/snippet.cpp.o
/usr/local/Cellar/cmake/3.22.2/bin/cmake -E __run_co_compile --iwyu="/usr/local/Cellar/include-what-you-use/0.17/libexec/bin/include-what-you-use;-Xiwyu;any;-Xiwyu;iwyu;-Xiwyu;args;-Xiwyu;--no_comments;-Xiwyu;--cxx17ns;-Xiwyu;max_line_length=120;--driver-mode=g++" -- /usr/local/Cellar/llvm/13.0.0_2/bin/clang++   -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk -std=gnu++2a -MD -MT CMakeFiles/test.dir/snippet.cpp.o -MF CMakeFiles/test.dir/snippet.cpp.o.d -o CMakeFiles/test.dir/snippet.cpp.o -c /Users/sebastien/Codes/Sandbox/IWYU/snippet.cpp
Error running '/usr/local/Cellar/include-what-you-use/0.17/libexec/bin/include-what-you-use': Segmentation fault
make[2]: *** [CMakeFiles/test.dir/build.make:76: CMakeFiles/test.dir/snippet.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/test.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
````

I have since my post upgraded my OS to macOS Monterey and the issue is still there (and IWYU 0.16 is perfectly fine).

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -1225,6 +1225,8 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // should be held responsible for a use.
   SourceLocation GetCanonicalUseLocation(SourceLocation use_loc,
                                          const NamedDecl* decl) {
+    CHECK_(decl != nullptr) << ": Need a decl to compute use location";
+
     // If we're not in a macro, just echo the use location.
     if (!use_loc.isMacroID())
       return use_loc;
@@ -2802,6 +2804,8 @@ class InstantiatedTemplateVisitor
   // instance, we're not responsible for a vector's call to
   // allocator::allocator(), because <vector> provides it for us).
   bool CanIgnoreDecl(const Decl* decl) const override {
+    if (!decl)
+      return true;
     return nodes_to_ignore_.Contains(decl);
   }
 

